pkg/volume/glusterfs/glusterfs.go

```diff
@@ -859,22 +872,28 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsPersi
 		return nil, 0, "", fmt.Errorf("failed to get cluster nodes for volume %s: %v", volume, err)
 	}
 
-	if len(p.provisionerConfig.customEpNamePrefix) == 0 {
-		epServiceName = string(p.options.PVC.UID)
-	} else {
-		epServiceName = p.provisionerConfig.customEpNamePrefix + "-" + string(p.options.PVC.UID)
+	addrlist := make([]v1.EndpointAddress, len(dynamicHostIps))
+	for i, v := range dynamicHostIps {
+		addrlist[i].IP = v
 	}
-	epNamespace := p.options.PVC.Namespace
-	endpoint, service, err := p.createEndpointService(epNamespace, epServiceName, dynamicHostIps, p.options.PVC)
+	subset := make([]v1.EndpointSubset, 1)
```

Comment:
is this re-entrant, or if the update fails, will we be left with a non-functional volume and service/endpoint?

Reply:
what happens today if the service/endpoint create fails? is the function re-entrant?

Reply:
@liggit if the update failed, we have to cleanup the volume and endpoint. I have updated the patchset with this cleanup operation. Thanks !!

Reply:
if this fails, does CreateVolume get recalled for the same PV later? if so, can it reuse the work already done, rather than trying to clean up every time? attempting to delete (like we did before with the volume, and like we're doing in this PR with the service/endpoint) does not mean we are re-entrant... network problems or process exit could leave us stranded before we have the ability to delete.

The three things this method is doing are:
* create volume
* create service
* create endpoint with volume IP

There are multiple possibilities for each of those operations:
* succeeds
* fails because it already exists
* fails for some other reason

If it already exists, are we responsible for reacting to the actual state of the already-existing thing? What does  that look like for each of the object types?

If it fails for some other reason, are we expected to fully clean up after ourselves (hint: that's not possible to guarantee... see comment about process exit or network errors), or simply return an error and get called again in the future to retry?

Reply:
@liggitt , previously the order was, 

- create volume
- create ep/svc 

In above workflow, we came across a limitation where eventhough the volume creation succeeded, svc creation failed due to quota setting in a customer setup . Considering volume creation and deletion is heavy compared to ep/svc creation this patch got introduced. Now, if the ep/svc creation failed or if volume creation failed , the cleanup is really light ie nothing but cleanup svc. I think the additional improvements what we can think of here is, can we make use of already created volume if there is a retry (ex:(  if ep update failed).  iic,  provision() from controller is on different go routine till it satisfies the PVC and the request is based on pvc UID. Unless user deleted and recreated the pvc object with same manifest the UID should remain same. But , the caveat here is that, heketi create a volume for each request with a different dynamically generated UID or volume name. To avoid that or to reuse what heketi already created in any of the previous iteration, we should map the volid between PVC UID and gluster/heketi vol ID ( somewhere) , such mapping does not exist. Also, as you mentioned, there are different combinations exist and we have to think through all of these. Till then, I think its better to track each transaction differently. I will also think through the possibilities of improving the workflow based on the scenarios you mentioned above, may be in a future followup design PR. :), does it sound fine ?